Packet++/header/SmtpLayer.h

```diff
@@ -0,0 +1,340 @@
+#ifndef PACKETPP_SMTP_LAYER
+#define PACKETPP_SMTP_LAYER
+
+#include "PayloadLayer.h"
+#include "SingleCommandTextProtocol.h"
+
+/// @file
+
+/**
+ * \namespace pcpp
+ * \brief The main namespace for the PcapPlusPlus lib
+ */
+namespace pcpp
+{
+	/**
+	 * Class for general SMTP message
+	 */
+	class SmtpLayer : public SingleCommandTextProtocol
+	{
+	  protected:
+		SmtpLayer(uint8_t *data, size_t dataLen, Layer *prevLayer, Packet *packet)
+			: SingleCommandTextProtocol(data, dataLen, prevLayer, packet)
+		{
+			m_Protocol = SMTP;
+		};
+
+		SmtpLayer(const std::string &command, const std::string &option) : SingleCommandTextProtocol(command, option)
+		{
+			m_Protocol = SMTP;
+		};
+
+	  public:
+		/**
+		 * A static method that checks whether the port is considered as SMTP control
+		 * @param[in] port The port number to be checked
+		 */
+		static bool isSmtpPort(uint16_t port) { return port == 25 || port == 587; }
+
+		// overridden methods
+
+		/// SMTP is the always last so does nothing for this layer
+		void parseNextLayer() {}
+
+		/**
+		 * @return Get the size of the layer
+		 */
+		size_t getHeaderLen() const { return m_DataLen; }
+
+		/// Does nothing for this layer
+		void computeCalculateFields() {}
+
+		/**
+		 * @return The OSI layer level of SMTP (Application Layer).
+		 */
+		OsiModelLayer getOsiModelLayer() const { return OsiModelApplicationLayer; }
+	};
+
+	/**
+	 * Class for representing the request messages of SMTP Layer
+	 */
+	class SmtpRequestLayer : public SmtpLayer
+	{
+	  public:
+		/**
+		 * Enum for SMTP command codes
+		 */
+		enum class SmtpCommand : int
+		{
+			/// Unknown command
+			UNK,
+			/// Starting mail body
+			DATA = ('D') | ('A' << 8) | ('T' << 16) | ('A' << 24),
+			/// Initiate conversation
+			EHLO = ('E') | ('H' << 8) | ('L' << 16) | ('O' << 24),
+			/// Expand the mailing list
+			EXPN = ('E') | ('X' << 8) | ('P' << 16) | ('N' << 24),
+			/// Initiate conversation
+			HELO = ('H') | ('E' << 8) | ('L' << 16) | ('O' << 24),
+			/// Ask information
+			HELP = ('H') | ('E' << 8) | ('L' << 16) | ('P' << 24),
+			/// Sender indication
+			MAIL = ('M') | ('A' << 8) | ('I' << 16) | ('L' << 24),
+			/// No operation
+			NOOP = ('N') | ('O' << 8) | ('O' << 16) | ('P' << 24),
+			/// Close conversation
+			QUIT = ('Q') | ('U' << 8) | ('I' << 16) | ('T' << 24),
+			/// Receiver indication
+			RCPT = ('R') | ('C' << 8) | ('P' << 16) | ('T' << 24),
+			/// Abort transaction
+			RSET = ('R') | ('S' << 8) | ('E' << 16) | ('T' << 24),
+			/// Identify user
+			VRFY = ('V') | ('R' << 8) | ('F' << 16) | ('Y' << 24),
+			/// Start TLS handshake
+			STARTTLS = ('S') | ('T' << 8) | ('L' << 16) | ('S' << 24),
```

Comment:
There is bug for STARTTLS packets since the command is longer than four characters. I'll fix it.